src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,110 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        this.elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int initialCapacity) {
+        this.elements = (T[]) new Object[initialCapacity];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elements.length) {
+            elements = grow(size);
+        }
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckForAdd(index);
+        if (size == elements.length) {
+            elements = grow(size);
+        }
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        if (size < elements.length) {
+            elements = grow(size + 1);
+        }
+        System.arraycopy(elements, size, elements, size, size - 1);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T deleteElement = elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - 1 - index);
+        size = size - 1;
+        return deleteElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((elements[i] == element) || (elements[i] != null
+                    && elements[i].equals(element))) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Couldn't find element " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void rangeCheckForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds");
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds");
+        }
+    }
+
+    private T[] grow(int oldCapacity) {
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
```

Comment:
```suggestion
    private T[] grow() {
        int newCapacity = elements.length + (elements.length >> 1);
```
you should increase capacity when size == elements.length, so why passing 'size' here
You already have oldCapacity - it's elements.length